Story/document upload in bestandsdelen

Rick-Ritense opened the pull request:
<!-- Please consider the following points before creating a pull request. -->

### Describe the changes
Link to the related Github issue: 

<!-- Please add any additional comments that might be relevant for reviewing this pull request  -->
<!-- Why did you choose to make these changes? Were there any trade-offs you had to consider?   -->
<!-- Note: add an empty line with a > to use multiple lines  -->
**Relevant comments:**
In Amsterdam there is a hard limit of 2 MB on POST calls which makes it impossible to upload files using the current document api method.

The documenten api also supports a "Bestandsdelen" method that allows a PUT call and uploading in chunks. This PR adds the new functionality
> 

### Breaking changes
<!-- Valtimo aims to comply with the SemVer specification.  -->
<!-- Breaking changes are only allowed in the `next-major` branch.  -->
- [x] The contribution only contains changes that are not breaking.

### Documentation
<!-- Release notes should be available in the Valtimo documentation.  -->
- [x] Release notes have been written for these changes. 

Link to the pull request in the [Valtimo documentation repository](https://github.com/valtimo-platform/valtimo-documentation):
> https://github.com/valtimo-platform/valtimo-documentation/pull/750

New features or changes that have been introduced have been documented.
- [x] Yes
- [ ] Not applicable

### Tests
Unit tests have been added that cover these changes
- [x] Yes
- [ ] Not applicable

Integration tests have been added that cover these changes
- [x] Yes
- [ ] Not applicable

### Security
The [Secure by Default principle](https://en.wikipedia.org/wiki/Secure_by_default) has been applied to these changes
- [ ] Yes
- [x] Not applicable

Added or changed REST API endpoints have authentication and authorization in place
- [ ] Yes
- [x] Not applicable

Valtimo access control checks have been implemented
- [ ] Yes
- [x] Not applicable

### Dependencies
Newly added dependencies do not introduce known vulnerabilities/CVE's and are in line with the Valtimo license
- [ ] Yes
- [x] Not applicable

Commit:
feature document upload in bestandsdelen

Initial commit

The code should be complete like this but I get an error during the put to bestandsdelen that the lock parameter is required, but it is provided with a filled value.

Don't know yet what the issue is

Commit:
Merge branch 'refs/heads/next-minor' into feature/document-upload-in-bestandsdelen

Commit:
Merge branch 'refs/heads/next-minor' into feature/document-upload-in-bestandsdelen

Commit:
feature - document upload in parts

Tried something new, still doesn't work

Commit:
Merge branch 'next-minor' into feature/document-upload-in-bestandsdelen

# Conflicts:
#	zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/DocumentenApiPlugin.kt
#	zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/DocumentenApiClient.kt

Commit:
Merge branch 'next-minor' into feature/document-upload-in-bestandsdelen

Commit:
feature document upload in bestandsdelen

Switched over to RestClient, got it working

Commit:
feature document upload in bestandsdelen

Switched over to RestClient, got it working

Commit:
feature document upload in bestandsdelen

Now getting the lock from the bestandsdelen array and added tests

Rick-Ritense committed:
Merge branch 'next-minor' into story/document-upload-in-bestandsdelen

Commit:
feature document upload in bestandsdelen

Added missing license headers

Commit:
feature document upload in bestandsdelen

Now using the bestandsdelen to determine the amount and size of the chunks

Commit:
Merge remote-tracking branch 'origin/story/document-upload-in-bestandsdelen' into story/document-upload-in-bestandsdelen

Rick-Ritense commented:
There is one problem at the moment. When you use a smaller chunk size (environment variable DOCUMENTEN_UPLOAD_CHUNK_SIZE for openzaak) the last (smaller) chunk is not accepted by the documenten api

Commit:
feature document upload in bestandsdelen

Now using the correct bestandsdeel UUID for each chunk

Commit:
Fixed failing tests

Also fixed some sonarlint findings

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/document-upload-in-bestandsdelen

Commit:
Fixed import

IvarKoreman-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/DocumentenApiClient.kt:
```diff
@@ -72,6 +77,75 @@ class DocumentenApiClient(
+         // Inside the CreateDocumentResult there is an array of bestandsdelen.
+         // Each bestandsdeel needs to be sent separately
+         // So the documenten api determines the amount of chunks, not this application.
+         val fileBytes = request.inhoud.readAllBytes()
```
This can be problematic. This loads the entire file into memory, so if the file is too large (or too many files are uploaded at the same time), we'll get an OutOfMemoryError.

Rick-Ritense replied:
What would be the solution to this? The documenten api expects a binary not a stream

IvarKoreman-Ritense replied:
`request.inhoud` is an `InputStream`. Instead of reading all bytes you can just read the ones you need to send.

Inside of the for-loop you could use something like this:

```
val chunk = ByteArray(bestandsdeel.omvang)
request.inhoud.read(chunk)
```

After that, chunk will contain the data that is read from the InputStream.

This does assume that that the sum of all `bestandsdeel.omgvang` add up to the total size of bytes that can be read (same as the code does now). The `InputStream.read(byte[])` method returns the number of bytes read, so it could be wise to assert that the return value is equal to the `chunk.size`. Also at the end you can use the `.read()` method of the InputStream [to see if it returns `-1`](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/io/InputStream.html#read()) to ensure that the end of the stream has been reached.

IvarKoreman-Ritense committed:
Reading only the bytes that will be sent for that chunk instead of the file as a whole.

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/DocumentenApiPlugin.kt:
```diff
@@ -295,6 +324,59 @@ class DocumentenApiPlugin(
+        metadata: Map<String, Any?>,
+        bestandsnaam: String?,
+        inhoudAsInputStream: InputStream,
+    ):CreateDocumentResult {
```
indent missing

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/DocumentenApiPlugin.kt:
```diff
@@ -295,6 +324,59 @@ class DocumentenApiPlugin(
+    private fun storeDocumentInParts(
+        execution: DelegateExecution,
+        metadata: Map<String, Any?>,
+        bestandsnaam: String?,
```
Why optional?

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/DocumentenApiPlugin.kt:
```diff
@@ -295,6 +324,59 @@ class DocumentenApiPlugin(
+     */
+    private fun storeDocumentInParts(
+        execution: DelegateExecution,
+        metadata: Map<String, Any?>,
```
Any should not be optional

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/CreateDocumentResult.kt:
```diff
@@ -24,4 +24,24 @@ class CreateDocumentResult(
     val beginRegistratie: LocalDateTime,
-)
\ No newline at end of file
+    val bestandsdelen: List<Bestandsdelen>?
```
optioneel?

Rick-Ritense replied:
Yes it is only returned when the inhoud parameter is kept empty in the request. 

As stated in the api description https://vng-realisatie.github.io/gemma-zaken/standaard/documenten/index 

bestandsdeel has a 0..* relation with enkelvoudiginformatieobject

Rick-Ritense replied:
Oh I now see that it also required in the response .... lol that sounds like a mismatch, but I will make it a mandatory parameter then :)

ivo-ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/Bestandsdelen.kt:
```diff
@@ -0,0 +1,25 @@
+
+package com.ritense.documentenapi.client
+
+class Bestandsdelen(
```
Why is this plural?

Tom-Ritense replied:
Why not single bestandeel

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/DocumentenApiClient.kt:
```diff
@@ -72,6 +77,78 @@ class DocumentenApiClient(
+        baseUrl: URI,
+        request: BestandsdelenRequest,
+        createDocumentResult: CreateDocumentResult,
+        bestandsnaam: String?
```
optional?

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/CreateDocumentResult.kt:
```diff
@@ -24,4 +24,24 @@ class CreateDocumentResult(
+        return bestandsdelen[0].url.substring(bestandsdelen[0].url.lastIndexOf("/") + 1)
+    }
+
+    fun getLockFromBestanddelen(): String {
```
not getLockFromBestand**s**Delen?

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/DocumentenApiClient.kt:
```diff
@@ -72,6 +77,78 @@ class DocumentenApiClient(
+         // So the documenten api determines the amount of chunks, not this application.
+         var response: BestandsdelenResult? = null
+
+         logger.info("Starting upload of file in {} chunks\n", createDocumentResult.bestandsdelen?.size)
```
This logging style is not kotlin style pls do kotlin style

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/DocumentenApiClient.kt:
```diff
@@ -72,6 +77,78 @@ class DocumentenApiClient(
+
+         logger.info("Starting upload of file in {} chunks\n", createDocumentResult.bestandsdelen?.size)
+
+         for(bestandsdeel in createDocumentResult.bestandsdelen!!) {
```
createDocumentResult.bestandsdelen.forEach {} instead

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/DocumentenApiClient.kt:
```diff
@@ -72,6 +77,78 @@ class DocumentenApiClient(
+         logger.info("Starting upload of file in {} chunks\n", createDocumentResult.bestandsdelen?.size)
+
+         for(bestandsdeel in createDocumentResult.bestandsdelen!!) {
+             logger.info("Sending chunk #{} for a size of {}",
```
idem

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/DocumentenApiClient.kt:
```diff
@@ -72,6 +77,78 @@ class DocumentenApiClient(
+                 }
+             }
+
+             val body: MultiValueMap<String, Any> = LinkedMultiValueMap()
```
Can this be moved to a dedicated data class ? with the fileResource also?

ivo-ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/DocumentenApiClient.kt:
```diff
@@ -72,6 +77,78 @@ class DocumentenApiClient(
+         // So the documenten api determines the amount of chunks, not this application.
+         var response: BestandsdelenResult? = null
+
+         logger.info("Starting upload of file in {} chunks\n", createDocumentResult.bestandsdelen?.size)
```
```suggestion
         logger.info { "Starting upload of file $bestandsnaam in ${createDocumentResult.bestandsdelen?.size} chunks" }
```

Rick-Ritense replied:
I will change the log and I will check the other calls. Is info the correct level or is debug better?

ivo-ritense replied:
It's hard to say. In general we want to log changes to external systems as info, but we also don't want an overload of information. If anything this one is probably fine, and the log line per chunk might have to be debug. @marijnritense @Klaas-Ritense what do you think?

Rick-Ritense replied:
I made the first log line info and the line for each chunk (and size) debug

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/DocumentenApiClient.kt:
```diff
@@ -72,6 +77,78 @@ class DocumentenApiClient(
+         return response ?: throw IllegalStateException("Upload in chunks failed")
+    }
+
+    fun unlockDocument(authentication: DocumentenApiAuthentication, baseUrl: URI,
```
Each param on a single line indentend pls

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/DocumentenApiClient.kt:
```diff
@@ -72,6 +77,78 @@ class DocumentenApiClient(
+
+    fun unlockDocument(authentication: DocumentenApiAuthentication, baseUrl: URI,
+                       documentLock: DocumentLock, bestandsdelenId: String) {
+        restClient(authentication)
```
Why not use existing method unlockInformatieObject

Rick-Ritense replied:
I missed that there already was a method for this. I will switch to the existing version

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/DocumentenApiClient.kt:
```diff
@@ -297,5 +371,6 @@ class DocumentenApiClient(
 
     companion object {
         const val ITEMS_PER_PAGE = 100
+        var logger = KotlinLogging.logger{}
```
val instead of var

Tom-Ritense commented on zgw/documenten-api/src/test/kotlin/com/ritense/documentenapi/client/DocumentenApiClientTest.kt:
```diff
@@ -59,9 +59,11 @@ import org.springframework.web.reactive.function.client.ClientRequest
 import java.net.URI
 import java.time.LocalDate
 import java.time.LocalDateTime
+import java.util.*
```
no * imports are allowed

Rick-Ritense replied:
```suggestion
import java.util.UUID
```

Tom-Ritense commented on zgw/documenten-api/src/test/kotlin/com/ritense/documentenapi/service/DocumentenApiServiceTest.kt:
```diff
@@ -28,30 +27,23 @@ import com.ritense.documentenapi.client.DocumentInformatieObject
 import java.time.LocalDate
 import java.time.LocalDateTime
-import java.util.UUID
+import java.util.*
```
idem

ivo-ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/DocumentenApiClient.kt:
```diff
@@ -72,6 +77,78 @@ class DocumentenApiClient(
+             "Failed to upload the full file. The file is larger than the sum of the omvang of all bestandsdelen."
+         }
+
+         return response ?: throw IllegalStateException("Upload in chunks failed")
```
This only returns the response of a single chunk upload. how valuable is that?

Rick-Ritense replied:
The response is not used for anything I only used it for testing. Cleaning it up

ivo-ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/DocumentenApiClient.kt:
```diff
@@ -107,12 +185,8 @@ class DocumentenApiClient(
-        if (documentSearchRequest.zaakUrl == null) {
-            throw IllegalArgumentException("Zaak URL is required")
-        }
+        require(ITEMS_PER_PAGE % pageable. pageSize == 0) { "Page size is not supported" }
```
```suggestion
        require(ITEMS_PER_PAGE % pageable.pageSize == 0) { "Page size is not supported" }
```

ivo-ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/BestandsdelenRequest.kt:
```diff
@@ -0,0 +1,24 @@
+
+class BestandsdelenRequest(
+    val inhoud: InputStream,
+    val lock:String
```
```suggestion
    val lock: String
```

ivo-ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/CreateDocumentResult.kt:
```diff
@@ -24,4 +24,24 @@ class CreateDocumentResult(
\ No newline at end of file
+    val bestandsdelen: List<Bestandsdelen>?
+) {
+    fun getBestandsdelenIdFromUrl(): String {
```
Where is this used? Only in tests? 


Is this the id of the bestandsdeel? If so, it's strange to always pick the first one from an array.

Rick-Ritense replied:
I think this is a leftover from an old attempt. I will clean it up

ivo-ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/CreateDocumentResult.kt:
```diff
@@ -24,4 +24,24 @@ class CreateDocumentResult(
+            return ""
+        }
+
+        return bestandsdelen[0].lock
```
Does the spec say the lock is always the same for every bestandsdeel?

Rick-Ritense replied:
It is a lock of the related informatieobject, not of the bestandsdeel. So yes it is the same for every bestandsdeel

lock required | string (lock)
Hash string, which represents id of the lock of related informatieobject

Rick-Ritense replied:
I will debug the application for a definitive answer

Rick-Ritense replied:
<img width="824" alt="image" src="https://github.com/user-attachments/assets/dd7b62a5-0bd6-4d97-bacb-6febba310b70">

Yes it is the same lock value for each bestandsdeel

Rick-Ritense committed:
Update zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/DocumentenApiClient.kt

Co-authored-by: Ivo Zaal <74657121+ivo-ritense@users.noreply.github.com>

Commit:
Code review findings

Commit:
Merge remote-tracking branch 'origin/story/document-upload-in-bestandsdelen' into story/document-upload-in-bestandsdelen

Commit:
Code review findings

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/document-upload-in-bestandsdelen

Commit:
Fixed a failing test

After bestandsdelen got required the DocumentenApiPluginIT tests failed

Commit:
Merge remote-tracking branch 'origin/story/document-upload-in-bestandsdelen' into story/document-upload-in-bestandsdelen

IvarKoreman-Ritense committed:
Replaced wildcard imports.

Commit:
Code review findings

Created a new data class: FileUploadPart including a test class. Moved the file upload part creation to this dataclass

Commit:
Merge remote-tracking branch 'origin/story/document-upload-in-bestandsdelen' into story/document-upload-in-bestandsdelen

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/DocumentenApiPlugin.kt:
```diff
@@ -146,8 +149,34 @@ class DocumentenApiPlugin(
             beschrijving = null,
             informatieobjecttype = null,
-            storedDocumentUrl = DOCUMENT_URL_PROCESS_VAR,
+            storedDocumentUrl = "",
```
Why empty string?

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/Bestandsdeel.kt:
```diff
@@ -0,0 +1,25 @@
+
+package com.ritense.documentenapi.client
+
+class Bestandsdeel(
```
Why not a data class?

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/BestandsdelenRequest.kt:
```diff
@@ -0,0 +1,24 @@
+
+import java.io.InputStream
+
+class BestandsdelenRequest(
```
Idem on data class

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/BestandsdelenResult.kt:
```diff
@@ -0,0 +1,24 @@
+ * limitations under the License.
+ */
+
+class BestandsdelenResult(
```
Idem on data class

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/DocumentenApiClient.kt:
```diff
@@ -72,6 +75,49 @@ class DocumentenApiClient(
+             }
+
+             val uploadPart = FileUploadPart(chunk, bestandsnaam, request.lock)
+             val body = uploadPart.createBody()
```
Single use of var is not clean

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/DocumentenApiClient.kt:
```diff
@@ -72,6 +75,49 @@ class DocumentenApiClient(
+         createDocumentResult.bestandsdelen.forEach { bestandsdeel ->
+             logger.debug { "Sending chunk #${bestandsdeel.volgnummer} for a size of ${bestandsdeel.omvang} bytes" }
+
+             val chunk = ByteArray(bestandsdeel.omvang)
```
This complete piece of code can move intp FileUploadPart if you ask me.

`
val chunk = ByteArray(bestandsdeel.omvang)
val bytesRead = request.inhoud.read(chunk)

check (bytesRead == chunk.size) {
                 "Failed to read all the bytes to upload. The expected omvang is larger than the bytes available."
}
`

Commit:
Code review findings

Moved more functionality to the data class FileUploadPart

Transformed classes to data classes

Reverted a change that was not needed

Updated tests

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/document-upload-in-bestandsdelen

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/DocumentenApiPlugin.kt:
```diff
@@ -295,6 +324,58 @@ class DocumentenApiPlugin(
+        client.unlockInformatieObject(
+            authenticationPluginConfiguration,
+            URI.create(documentCreateResult.url),
+            documentLock)
```
indent is wrong

Tom-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/client/DocumentenApiClient.kt:
```diff
@@ -297,5 +333,6 @@ class DocumentenApiClient(
 
     companion object {
         const val ITEMS_PER_PAGE = 100
+        val logger = KotlinLogging.logger{}
```
space is needed between loggern and {} pls run the intellj formatter.

Commit:
Ran IntelliJ code formatter on changed files

Commit:
Merge remote-tracking branch 'origin/story/document-upload-in-bestandsdelen' into story/document-upload-in-bestandsdelen

ivo-ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/DocumentenApiPlugin.kt:
```diff
@@ -53,7 +56,7 @@ import org.springframework.web.util.UriComponentsBuilder
 import java.net.URI
 import java.time.LocalDate
-import java.util.UUID
+import java.util.*
```
```suggestion
import java.util.UUID
```

ivo-ritense committed:
Update zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/DocumentenApiPlugin.kt

ivo-ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/DocumentenApiPlugin.kt:
```diff
@@ -148,6 +151,32 @@ class DocumentenApiPlugin(
             storedDocumentUrl = DOCUMENT_URL_PROCESS_VAR,
         )
+
+        execution.setVariable(DOCUMENT_URL_PROCESS_VAR, documentCreateResult.url)
```
```suggestion
```

ivo-ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/DocumentenApiPlugin.kt:
```diff
@@ -295,6 +324,60 @@ class DocumentenApiPlugin(
+            inhoudAsInputStream = InputStream.nullInputStream(),
+            beschrijving = null,
+            informatieobjecttype = null,
+            storedDocumentUrl = ""
```
Will this work?

ivo-ritense committed:
Update zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/DocumentenApiPlugin.kt

ivo-ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/DocumentenApiPlugin.kt:
```diff
@@ -295,6 +323,60 @@ class DocumentenApiPlugin(
+            documentLock
+        )
+
+        return documentCreateResult
```
```suggestion
```

ivo-ritense committed:
Update zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/DocumentenApiPlugin.kt

ivo-ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/DocumentenApiPlugin.kt:
```diff
@@ -295,6 +323,59 @@ class DocumentenApiPlugin(
+        metadata: Map<String, Any>,
+        bestandsnaam: String,
+        inhoudAsInputStream: InputStream,
+    ): CreateDocumentResult {
```
```suggestion
    ) {
```

ivo-ritense committed:
Update zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/DocumentenApiPlugin.kt

ivo-ritense commented on zgw/documenten-api/src/test/kotlin/com/ritense/documentenapi/DocumentenApiPluginTest.kt:
```diff
@@ -307,7 +311,7 @@ internal class DocumentenApiPluginTest {
         val apiRequestCaptor = argumentCaptor<CreateDocumentRequest>()
         verify(client).storeDocument(any(), any(), apiRequestCaptor.capture())
-        verify(executionMock).setVariable(DOCUMENT_URL_PROCESS_VAR, "returnedUrl")
+        verify(executionMock, times(2)).setVariable(DOCUMENT_URL_PROCESS_VAR, "returnedUrl")
```
```suggestion
        verify(executionMock).setVariable(DOCUMENT_URL_PROCESS_VAR, "returnedUrl")
```

ivo-ritense committed:
Update zgw/documenten-api/src/test/kotlin/com/ritense/documentenapi/DocumentenApiPluginTest.kt

ivo-ritense commented on zgw/documenten-api/src/test/kotlin/com/ritense/documentenapi/DocumentenApiPluginTest.kt:
```diff
@@ -229,7 +233,7 @@ internal class DocumentenApiPluginTest {
         val apiRequestCaptor = argumentCaptor<CreateDocumentRequest>()
         verify(client).storeDocument(any(), any(), apiRequestCaptor.capture())
-        verify(executionMock).setVariable(DOCUMENT_URL_PROCESS_VAR, "returnedUrl")
+        verify(executionMock, times(2)).setVariable(DOCUMENT_URL_PROCESS_VAR, "returnedUrl")
```
Check if this should be 1x

Commit:
Fixed the last test

Rick-Ritense committed:
Update zgw/documenten-api/src/test/kotlin/com/ritense/documentenapi/client/DocumentenApiClientTest.kt
